community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java

```diff
@@ -490,6 +490,18 @@ public void start() throws IOException
         // Monitor listeners
         LoggingLogFileMonitor loggingLogMonitor = new LoggingLogFileMonitor( logging.getMessagesLog( getClass() ) );
         monitors.addMonitorListener( loggingLogMonitor );
+        monitors.addMonitorListener( new Recovery.MonitorAdapter()
+        {
+            @Override
+            public void recoveryRequired( long recoveredLogVersion )
+            {
+                // This method will be called before recovery actually starts and so will ensure that
+                // each store is aware that recovery will be performed. At this point all the stores have
+                // already started btw.
+                // Go and read more at {@link CommonAbstractStore#recoveryRequired()}
+                neoStoreModule.neoStore().recoveryIsRequired();
+            }
+        } );
```

Comment:
I don't like the idea of injecting important behaviour through monitors. Could this not be put in the SPI.getRecoverer implementation in NeoStoreDataSource?